### ISSUE TITLE
CTRL+C (3) not returned by getch() on WinCon and WinGUI

### ISSUE THREAD

**Issue by rhaberkorn** [Bill-Gray/PDCursesMod#226]
In this small example, CTRL+C will never be reported (as 3) by getch():

```c
#include <curses.h>

int main(int argc, char **argv)
{
	int key;

	initscr();

#if 0
	for (int i = 0; i < PDC_MAX_FUNCTION_KEYS; i++)
		PDC_set_function_key(i, 0);
#endif

	/* NOTE: Order of cbreak() and raw() is significant on ncurses */
	cbreak();
	raw();
	noecho();

	do {
		key = getch();	
		printw("Key %d\n", key);
	} while (key != 'q');

	endwin();
	return 0;
}
```

Pressing CTRL+C does not print anything (getch() does not return). It makes no difference whether we disable all function keys. It has been tested on v4.3.1 as installed on MinGW with both WinCon (`-lpdcurses_wincon`) as well as WinGUI (`-lpdcurses_wingui`). I know it used to work on PDCurses/win32 (original) at some point in the past.
The example works fine with ncurses on Linux.

Any clues? I am also open for workarounds in order to support the PDCurses currently installable with MinGW.

**Comment by Bill-Gray** [comment]
The problem is in [this commit](https://github.com/Bill-Gray/PDCursesMod/commit/9f487d4feaa15d250fb04abb8e16fe6ed36f9ae0#diff-1616304a056bb8840a48c5f5e3f4bba6e5731e7865e77b3e5d137445fa500dc7).  For Windows,  Ctrl-C now means 'copy',  and Ctrl-V means 'paste'.  For the nonce,  removing the `#ifndef _WIN32` and `#endif` statements from that commit (i.e.,  reverting it) will get you those keys back.

Longer term,  it seems to me that `PDC_set_function_key()` ought to be used here.  Then,  by default,  Ctrl-C and Ctrl-V would still be used for copy/paste (and Shift-Ctrl-C and Shift-Ctrl-V on non-Windows platforms).

But if you actually wanted your program to receive those keys,  you'd call
```
PDC_set_function_key( FUNCTION_KEY_COPY, 0);
PDC_set_function_key( FUNCTION_KEY_PASTE, 0);
```
(which would disable copy/paste;  you might just assign other keys to those functions.)

Also of note : Ctrl-C isn't returned on the VT and framebuffer port,  either,  because it gets caught by the signal handler rather than simply being processed as a "normal" key.  Same problem as you report,  but with a different cause.  This should be an easy fix,  though.

**Comment by rhaberkorn** [comment]
I would have picked up the CTRL+V question as I also noticed it. CTRL+V however is not as important as CTRL+C for [my application](https://github.com/rhaberkorn/sciteco/).
`FUNCTION_KEY_PASTE` already exists but is not referenced in the code base at all.

> For the nonce, removing the #ifndef _WIN32 and #endif statements from that commit (i.e., reverting it) will get you those keys back.

Is there any workaround that would work with v4.3.1 without recompilation? I might also get away with downgrading to v4.2.0 for the time being.

> Also of note : Ctrl-C isn't returned on the VT and framebuffer port...

Off topic, but what's the motivation for the VT and framebuffer ports? When I required a small VT curses for obscure platforms, I would usually link in [netbsd-curses](https://github.com/sabotage-linux/netbsd-curses) as it has minimal runtime requirements.
And a Linux framebuffer implementation... Why not simply run a framebuffer terminal emulator with ncurses?

**Comment by Bill-Gray** [comment]
Correct,  `FUNCTION_KEY_PASTE` really ought to have been used here.  I am working on a revision that will make this run sanely (and it'll also clean up the code quite a bit).

There's no workaround I can think of that wouldn't involve recompilation.  You could go with v4.2.0 for the nonce,  then switch to v4.2.2 when that comes out.

Truthfully,  I didn't know there was a port of netbsd-curses to Linux.  I'll have to investigate;  we may be able to help each other out,  and/or I may pick some bits and pieces out of them.  (It does not appear they are aware of PDCursesMod,  either.)

However,  even had I known,  netbsd-curses != PDCursesMod,  and I'd have proceeded with the VT and framebuffer ports.  They can be used by any PDCursesMod and most PDCurses programs with simple recompilation (or,  if you're using shared libraries,  by swapping in a different library.)    The framebuffer port enables basically all of PDCurses : 1048576 color pairs,  a full color palette,  under/over/left/right/strikeout lining,  and so forth.  The font can be changed on the fly with a hotkey.  (ncurses on the Linux console is quite limited : 16 colors,  which you can't change.)

The VT port is somewhat simpler (no terminfo concept,  for example),  at the price of leaving out some more unusual terminals.  The runtime requirements for both ports are quite low,  considerably less than for ncurses;  they're about as bare-bones as you can get.

**Comment by rhaberkorn** [comment]
> There's no workaround I can think of that wouldn't involve recompilation. You could go with v4.2.0 for the nonce, then switch to v4.2.2 when that comes out.

There was unfortunately no v4.2.0 in MinGW. But since @GitMensch seems to be very active in the MinGW packages, I hope that the fix will be rolled out soon to MinGW as well.

**Comment by Bill-Gray** [comment]
FWIW,  I've just patched the code to fix this.  Its main practical effect at present are that Ctrl-C is returned in WinGUI and WinCon.  Longer range,  it should enable the function keys in a more systematic way.  For example,  at present,  Alt-Minus cycles the font in the framebuffer port (if multiple fonts have been selected;  see the [`README.md` for that port](https://github.com/Bill-Gray/PDCursesMod/tree/master/fb#readme)).  That function key ought to be programmatically changeable.  But that's a problem for another day.

**Comment by GitMensch** [comment]
@Bill-Gray Do you mean PDCursesMod "4.3.1 has an issue that will be fixed with 4.3.2" - or do you really intend to create a bugfix version for 4.2?
In any case 4.3.1+ https://github.com/Bill-Gray/PDCursesMod/compare/v4.3.1...master#diff-03357ad51eea15a57e6a6253f0718dd13801d94a8321952b7df9d6c8d0b2897c is already quite nice with speedups and fixes, not sure if and what you want to add for an updated version. As soon as a new release is done it should be no big issue to update the MSYS2 package, as the build system did not change that much - mostly removing a patch and updating the source. My 5ct would be: possibly adjust the initscr name mangling, if the config-stuff isn't ready "sooner" then do a release without it, otherwise wait.
@rhaberkorn: if that's very important to you, you could apply the necessary changes for this fix to the MSYS2 build (important patches can be included to fix issues in published versions). I'd review but not do them.

**Comment by Bill-Gray** [comment]
@GitMensch I see I erred there;  I saw '4.2.n' and thought '4.3.n'.

4.2 was a while back,  and much has been added and fixed since then.  If you look at the history,  4.3 and 4.3.1 and changes since have fixed many bugs,  and we haven't tried to go back to bug-fix older versions.

Our better move might indeed be to proceed toward a 4.3.2,  one that has the Ctrl-C working in WinGUI and WinCon (not VT yet,  but that can wait until after the release) and that mangles `endwin()` instead of `initscr()`... both of which we now have;  see commit 1e19a2eb96 for the latter.

If this all works with MinGW,  and nobody suggests something else that Ought To Be Done,  I'll bump up `PDC_VER_CHANGE` from 1 to 2,  modify the `PDC_VER_YEAR/MONTH/DATE` #defines,   change the top line of `HISTORY.md` from "current state" to say it's 4.3.2,  and make it a for-real release... anyone see a reason not to do that?

**Comment by GitMensch** [comment]
If you find some time and could investigate the doc reorg from William it could be reasonable to do similar adjustments here, if not already done.

As all CI builds passed fine, I'd say - no reason to not do a release now. If the config stuff is not ready yet go with the current version.

**Comment by Bill-Gray** [comment]
Having heard no objections,  I've released 4.3.2,  so we should be good now,  and will close this issue.

I am in no rush to apply William's doc reorg.  I don't think it's particularly misguided;  I just don't see much point to it.  (Except that I do keep some "irrelevant" bits of PDCurses in sync which PDCursesMod just for the sake of making it easier to keep "relevant" bits in sync,  and that may apply here.)
